File: Advance/17.Implement Floyd-Warshall Algorithm for all-pairs shortest paths/Solution.cpp
#include <iostream>
#include <vector>
#include <limits>
using namespace std;

const int INF = 1e9; // A large number to represent "infinity"

void floydWarshall(vector<vector<int>> &dist, int V) {
    // Use each vertex as an intermediate point.
    for (int k = 0; k < V; k++) {
        for (int i = 0; i < V; i++) {
            for (int j = 0; j < V; j++) {
                if (dist[i][k] != INF && dist[k][j] != INF &&
                    dist[i][j] > dist[i][k] + dist[k][j]) {
                    dist[i][j] = dist[i][k] + dist[k][j];
                }
            }
        }
    }
}

int main() {
    int V, E;
    cout << "Enter the number of vertices: ";
    cin >> V;
    cout << "Enter the number of edges: ";
    cin >> E;
    
    // Initialize the graph as a V x V matrix.
    vector<vector<int>> dist(V, vector<int>(V, INF));
    for (int i = 0; i < V; i++) {
        dist[i][i] = 0;
    }
    
    cout << "Enter each edge in the format: u v weight\n";
    for (int i = 0; i < E; i++) {
        int u, v, w;
        cin >> u >> v >> w;
        // Assuming 0-indexed vertices.
        dist[u][v] = w;
        // For directed graph, comment out the following line.
        // dist[v][u] = w; 
    }
    
    floydWarshall(dist, V);
    
    cout << "\nShortest distances between every pair of vertices:\n";
    for (int i = 0; i < V; i++) {
        for (int j = 0; j < V; j++) {
            if (dist[i][j] == INF)
                cout << "INF ";
            else
                cout << dist[i][j] << " ";
        }
        cout << endl;
    }
    
    return 0;
}

File: Advance/17.Implement Floyd-Warshall Algorithm for all-pairs shortest paths/Solution.py
def floyd_warshall(dist, V):
    # Compute all pairs shortest paths using Floyd-Warshall algorithm.
    for k in range(V):
        for i in range(V):
            for j in range(V):
                if dist[i][k] != INF and dist[k][j] != INF:
                    dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j])

if __name__ == "__main__":
    INF = 10**9  # A large number to represent "infinity"
    
    V = int(input("Enter the number of vertices: "))
    E = int(input("Enter the number of edges: "))
    
    # Initialize the graph as a V x V matrix.
    dist = [[INF] * V for _ in range(V)]
    for i in range(V):
        dist[i][i] = 0
    
    print("Enter each edge in the format: u v weight")
    for _ in range(E):
        u, v, w = map(int, input().split())
        # Assuming vertices are 0-indexed.
        dist[u][v] = w
        # If the graph is undirected, uncomment the following line:
        # dist[v][u] = w
    
    floyd_warshall(dist, V)
    
    print("\nShortest distances between every pair of vertices:")
    for i in range(V):
        for j in range(V):
            if dist[i][j] == INF:
                print("INF", end=" ")
            else:
                print(dist[i][j], end=" ")
        print()
